File: nhglib-core/src/io/github/voidzombie/nhglib/runtime/messaging/Messaging.java
package io.github.voidzombie.nhglib.runtime.messaging;

import io.reactivex.Observable;
import io.reactivex.subjects.PublishSubject;

/**
 * Created by Fausto Napoli on 08/12/2016.
 */
public class Messaging {
    private final PublishSubject<Message> messagePublisher;

    public Messaging() {
        messagePublisher = PublishSubject.create();
    }

    public void send(Message message) {
        messagePublisher.onNext(message);
    }

    public Observable<Message> get(String ... filters) {
        return messagePublisher.filter(message -> {
            boolean res = false;

            for (String filter : filters) {
                if (message.is(filter)) {
                    res = true;
                    break;
                }
            }

            return res;
        });
    }
}

File: nhglib-core/src/io/github/movementspeed/nhglib/core/messaging/Messaging.kt
package io.github.movementspeed.nhglib.core.messaging

import io.reactivex.Observable
import io.reactivex.functions.Predicate
import io.reactivex.subjects.PublishSubject

/**
 * Created by Fausto Napoli on 08/12/2016.
 */
class Messaging {
    private val messagePublisher: PublishSubject<Message>

    init {
        messagePublisher = PublishSubject.create()
    }

    fun send(message: Message) {
        messagePublisher.onNext(message)
    }

    operator fun get(vararg filters: String): Observable<Message> {
        return messagePublisher.filter { message ->
            var res = false

            for (filter in filters) {
                if (message.`is`(filter)) {
                    res = true
                    break
                }
            }

            res
        }
    }
}
